Sources/Commands/SwiftRunCommand.swift

```diff
@@ -284,9 +284,20 @@ public struct SwiftRunCommand: SwiftCommand {
     /// A safe wrapper of TSCBasic.exec.
     private func execute(path: String, args: [String]) throws -> Never {
         #if !os(Windows)
-        // On platforms other than Windows, signal(SIGINT, SIG_IGN) is used for handling SIGINT by DispatchSourceSignal,
-        // but this process is about to be replaced by exec, so SIG_IGN must be returned to default.
-        signal(SIGINT, SIG_DFL)
+        // Dispatch will disable almost all asynchronous signals on its worker threads, and this is called from `async`
+        // context. To correctly `exec` a freshly built binary, we will need to:
+        // 1. reset the signal masks
+        for i in 1..<NSIG {
+            signal(i, SIG_DFL)
+        }
+        var sig_set_all = sigset_t()
+        sigfillset(&sig_set_all)
+        sigprocmask(SIG_UNBLOCK, &sig_set_all, nil)
+        
+        // 2. close all file descriptors.
+        for i in 3..<getdtablesize() {
+            close(i)
```

Comment:
Do we have any open fds managed by SwiftPM at this point? My concern is that this might corrupt the state of underlying libraries like libc and cause other problems for the case when `exec` fails for some reason. I think closing fds is not a part of this function's responsibilities but should be handled by those fds owners using `O_CLOEXEC`.

Reply:
> Do we have any open fds managed by SwiftPM at this point?
I don't think we should, but I don't have any certainty our `swift-tools-support-core` or `swift-corelibs-foundation` APIs don't leak anything.

Reply:
I think the discussion point here is whether to prioritize potential state corruption after the exec fails or potential fd leak during the exec'ed program execution. IMO the former issue is more critical than the latter since we don't have a large number of open fd's here and emitting proper error diagnostics when a problem occurs is essential for further investigation.

Reply:
IIUC the argument that Johannes is making is that `bash` is doing it this way and it works fine for them, whatever state underlying libraries have, it'll be gone with `execv` call anyway https://github.com/bminor/bash/blob/f3b6bd19457e260b65d11f2712ec3da56cef463f/execute_cmd.c#L6162C1-L6173

Reply:
Yes, that's what UNIX programs are expected to do if they exec: close the file descriptors that they're not intending to pass on. LGTM

Reply:
Okay, that makes sense to me 👍